File: erlang/triangle/src/triangle.erl
-module(triangle).

-export([kind/3]).


kind(A, B, C) when A =< 0 orelse B =< 0 orelse C =< 0 ->
    {error, "all side lengths must be positive"};
kind(A, B, C) when A + B =< C orelse A + C =< B orelse B + C =< A ->
    {error, "side lengths violate triangle inequality"};
kind(A, A, A) -> equilateral;
kind(A, _B, A) -> isosceles;
kind(_B, A, A) -> isosceles;
kind(A, A, _B) -> isosceles;
kind(_A, _B, _C) -> scalene.

File: elixir/triangle/triangle.exs
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  defguardp inequality(a, b, c) when a < b + c and b < a + c and c < a + b
  defguardp positive(a, b, c) when a > 0 and b > 0 and c > 0

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: {:ok, kind} | {:error, String.t()}
  def kind(a, b, c) when not positive(a, b, c) do
    {:error, "all side lengths must be positive"}
  end

  def kind(a, b, c) when not inequality(a, b, c) do
    {:error, "side lengths violate triangle inequality"}
  end

  def kind(a, a, a) do
    {:ok, :equilateral}
  end

  def kind(a, a, _) do
    {:ok, :isosceles}
  end

  def kind(a, _, a) do
    {:ok, :isosceles}
  end

  def kind(_, a, a) do
    {:ok, :isosceles}
  end
  def kind(_, _, _) do
    {:ok, :scalene}
  end
end
